Make Competitions.id immutable

timhabermaas opened the pull request:
Letting the end user change primary keys was a rather unfortunate design
decision which led to some weird corner cases. Especially keeping
foreign keys in sync was painful. This PR fixes that and introduces
another column: `friendly_id` which is meant to be shown and edited by the
user. `friendly_id` is generated using the same rules as `id`.

Note: `id` is still generated before validation in order to not break
existing assumptions/code.

:bangbang: **Not meant to be merged before #407 and #319**. :bangbang:

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -39,9 +39,9 @@ def create
     else
-      # Show id errors under name, since we don't actually show an
+      # Show friendly id errors under name, since we don't actually show an
       # id field to the user, so they wouldn't see any id errors.
```
nitpick: `id field` is mentioned twice here, maybe it should also be changed to `friendly id` or `friendly_id`?

timhabermaas replied:
Yeah, fixed.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -115,7 +116,11 @@ def create_id_and_cell_name
         # By replacing accented chars with their ascii equivalents, and then
```
not your fault, but `By` looks like it should be lower cased here.

timhabermaas replied:
Fixed.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -115,7 +116,11 @@ def create_id_and_cell_name
+        self.id = safe_name_without_year[0...(MAX_FRIENDLY_ID_LENGTH - year.length)] + year
+      end
+      if friendly_id.blank?
+        safe_name_without_year = ActiveSupport::Inflector.transliterate(name_without_year).gsub(/[^a-z0-9]+/i, '')
```
Did you consider moving `safe_name_without_year` up outside of these `if` statements instead of copying it? I don't think performance is a real concern here.

timhabermaas replied:
I was being lazy, sorry. Fixed.

jfly commented:
Wow, this is awesome! :heart: :heart: :heart: deleting code.

Can you explain your comment "Note: id is still generated before validation in order to not break
existing assumptions/code."? I was hoping that in the fullness of time we'd change the `id` field to a boring autoincrementing integer. I think it might be a good idea to do that as part of these changes, because I'm sure we're going to confuse the Results team otherwise.

You're absolutely right that we need to push #332 and #319 through first. I imagine the plan is then for you to update this PR (#381) to make that new rails code use `friendly_id` instead of `id`?

Don't we still have a lot of [results administration code](https://github.com/cubing/worldcubeassociation.org/tree/master/webroot/results/admin) that refers to the `Competition.id` field? (in particular, I know there's the [database export code](https://github.com/cubing/worldcubeassociation.org/blob/master/webroot/results/admin/export_public.php), and people would be confused if competitions started showing up with old `id`s) If so, what do you think is the best way to handle that? My plan was always to wait until we were rid of all our PHP code before we started making sweeping database improvements like this (#91).

timhabermaas commented:
>Can you explain your comment "Note: id is still generated before validation in order to not break
existing assumptions/code."? I was hoping that in the fullness of time we'd change the `id` field to a boring autoincrementing integer. I think it might be a good idea to do that as part of these changes, because I'm sure we're going to confuse the Results team otherwise.

Changing it to an autoincrementing integer is my goal as well, but I have no idea which parts of the application depend on a String `id` (or even depend on the specific structure), so I'm not too keen on changing it right now. E.g. I missed this:

>Don't we still have a lot of results administration code that refers to the `Competition.id` field?

As long as it only references the `id` it shouldn't be a problem. If you can change `id`s of competitions through the PHP code, though... :fire:

> I imagine the plan is then for you to update this PR (#381) to make that new rails code use `friendly_id` instead of `id`?

Yes, that's my plan. I also want to switch to `friendly_id`s for the URLs, maybe using the `friendly_id` gem.

> people would be confused if competitions started showing up with stale ids

How are people going to notice this? The relations `*.competitionId <-> Competition.id` stay intact and even if people depend on the contents of `Competitions.id` I don't see how they'd be influenced by that change. Do some people keep a history of exports and somehow detect updated competition by `id` changes?

jfly commented:
> How are people going to notice this? The relations *.competitionId <-> Competition.id stay intact and even if people depend on the contents of Competitions.id I don't see how they'd be influenced by that change. Do some people keep a history of exports and somehow detect updated competition by id changes?

Ah, good points. I do still think this will confuse the Results team, but I think that's fine so long as we give them fair warning. I think the one place where this will be an issue is when external websites link to the WCA website for competition results. Their current code will use `Competition.id`.... okay I just realized you didn't change our `Competition` query code in `competitions_controller` to look for `friendly_id`, was that intentional? What about all the places where we generate links to competitions, shouldn't we be using `friendly_id` as well? I may be misunderstanding the scope of your changes here.

timhabermaas commented:
> okay I just realized you didn't change our `Competition` query code in `competitions_controller` to look for `friendly_id`, was that intentional?

Kind of. I didn't want to change too many stuff before #332 and #319 make it into `master`. I indeed plan on using `friendly_id` for the URLs, though.

> What about all the places where we generate links to competitions, shouldn't we be using `friendly_id` as well?

As long as the links look like `competition_path(competition)` (instead of `competition_path(competition.id)`) overriding `Competition#to_param` and returning `friendly_id` instead of `id` will take care of this automatically (Rails link helpers call `to_param` on the parameter).

The `friendly_id` gem would take care of both, btw:
* Making `find(params[:id])` look for both `id` or `friendly_id`
* Generating the URLs using `friendly_id`

jfly commented:
Awesome, that answers all my questions! Outside websites may link to us with the `id` when ideally they'd use the `friendly_id`, but that's fine, because we'll accept both `id` and `friendly_id`!

timhabermaas commented:
@jfly I've just rebased this branch. Could you take another look at it (I didn't change any code, just resolved some conflicts in `db/structure.sql`)?
I'd consider it mergeable as long as the concerns you had regarding the results team/PHP code aren't an issue.

pedrosino commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -90,9 +90,9 @@ def create
-      # id field to the user, so they wouldn't see any id errors.
-      @competition.errors[:name].concat(@competition.errors[:id])
+      # Show friendly id errors under name, since we don't actually show an
+      # friendly id field to the user, so they wouldn't see any friendly id errors.
```
[Jeremy mode]
Nitpick: "...who a friendly id..."
[/Jeremy mode]

timhabermaas replied:
Done: 94bf6ccc2a8de5b82dff6e9c87cef6719b095965

pedrosino commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -49,6 +49,7 @@ en:
         use_wca_registration: "I would like to use the WCA website for registration"
         guests_enabled: "Guests"
         receive_registration_emails: "Notify me by email when someone registers for this competition"
+        friendly_id: "Public facing ID"
```
Maybe something like "Public display ID"
(I'm not that sure about my suggestion, though)

timhabermaas replied:
I'm not sure either. The common technical term is ["slug"](https://en.wikipedia.org/wiki/Semantic_URL#Slug), but there are probably not many people who know what it means.

jfly commented on WcaOnRails/db/migrate/20160207125654_add_friendly_id_to_competitions.rb:
```diff
@@ -0,0 +1,12 @@
+class AddFriendlyIdToCompetitions < ActiveRecord::Migration
+  def up
+    add_column :Competitions, :friendly_id, :string, null: false, default: ""
```
Should we have a unique constraint on `friendly_id`?

pedrosino replied:
I guess so.
Em 31/03/2016 19:41, "Jeremy Fleischman" <notifications@github.com>
escreveu:

> In WcaOnRails/db/migrate/20160207125654_add_friendly_id_to_competitions.rb
> <https://github.com/cubing/worldcubeassociation.org/pull/381#discussion_r58139669>
> :
>
> > @@ -0,0 +1,12 @@
> > +class AddFriendlyIdToCompetitions < ActiveRecord::Migration
> > +  def up
> > +    add_column :Competitions, :friendly_id, :string, null: false, default: ""
>
> Should we have a unique constraint on friendly_id?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/cubing/worldcubeassociation.org/pull/381/files/e5c12db34efa24d63b37c5b6945cd530603f3180#r58139669>
>

timhabermaas replied:
Makes sense, thanks!

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -110,12 +111,16 @@ def create_id_and_cell_name
         # Generate competition id from name
-        # By replacing accented chars with their ascii equivalents, and then
+        # by replacing accented chars with their ascii equivalents, and then
         # removing everything that isn't a digit or a character.
```
Nit: This comment looks out of place now that you moved `safe_name_without_year` up. Maybe move the comment as well?

jfly commented:
Thanks for updating this, Tim! A couple of things:

- Did you update `def competition_params` to not allow the `:id` param anymore? See https://github.com/jfly/worldcubeassociation.org/blob/2ce4fbddbe7c6ea0bab5d68a169accaf8bdb2b13/WcaOnRails/app/controllers/competitions_controller.rb#L293.
- #407 *still* hasn't been merged up (due to the same performance issues). We could either wait to merge this change up until #407 is merged, or we could update our existing PHP code to be aware of this.
- Have you analyzed all calls to `Competition.find` and `Competition.find_by_id` to make sure they're doing what we want now?

timhabermaas commented:
1. Yes: https://github.com/cubing/worldcubeassociation.org/pull/381/files#diff-cc669b7b0cc8520d68707875f678c920L293
2. That's unfortunate. I'll look into #407 next week and hopefully can get performance down to an acceptable level. :pray: 
3. No, I haven't touched that code and everything (finders/URLs) is still based on `id`s. I'll add `friendly_id` which _should_ take care of these things automatically.

jfly commented:
Great, thanks Tim! I've been meaning to get around to #407, but there is so much going on now, I hardly get to code anymore. I won't work on #407 without talking to you first, so feel free to take it on!

timhabermaas commented:
@jfly I've used the friendly_id (which I renamed to `slug` because that seems to be the common term for it) in the URLs by overriding `to_param` in the `Competition` model. I've also updated the finders to use `.find_by_slug!` instead of `.find`.

Currently `c.slug == c.id` is true for all competitions and this might be confusing. We should probably consider changing the `id` to an integer or some other non-human-readable type.

So, if this looks good to you, we could merge it.

timhabermaas commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -81,6 +81,7 @@ en:
         registration_open: "Registration open"
         registration_close: "Registration close"
         remarks: "Remarks"
+        slug: "Public facing ID"
```
@viroulep What would be a good French translation for that field?

viroulep replied:
I think "ID affiché publiquement" is good!

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -94,11 +94,11 @@ def info
         attributes = competition_to_clone.attributes
-        # Don't clone id, name, cellName, and dates.
-        %w(id name cellName year month day endMonth endDay registration_open registration_close results_posted_at).each { |attribute| attributes.delete attribute }
+        # Don't clone id, slug, name, cellName, and dates.
```
I fear this comment has started to rot (see `results_posted_at`). How about we remove the comment, or change it to "remove fields we shouldn't clone"?

jfly commented:
Thanks for ressurecting this. Before we merge this up, we need to give the Results Team a heads up.

Just check if I understand correctly: it looks like now our URLs only accept a slug, and won't work if you give them an id? If we don't change this, I think we should change our database export script to export the slug as the id, otherwise websites that use our database export won't be able to link back to us.

What about all our tables that have `competition_id` or `competitionId` as a foreign key? Are they all safe from this change?

What happens when the Results Team uploads results from a competition whose slug != it's id? I think this depends on if we change our public export at all: right now, the WCA Workbook Assistant will generate JSON with the id, which would be the right thing to do (we want our Results to point to a competition's id, not its slug, right?)

This is actually pretty hard for me to reason about =( Right now, I'm feeling we should wait until we're ready to go the whole nine yards and change the id to an autoincrementing integer, sorry. What's the motivation for pushing this half solution live now, instead of waiting to go the distance?

timhabermaas commented:
>Just check if I understand correctly: it looks like now our URLs only accept a slug, and won't work if you give them an id? If we don't change this, I think we should change our database export script to export the slug as the id, otherwise websites that use our database export won't be able to link back to us.

This is correct. I'll change it such that the URLs will work with ids as well in order to not break people's code.

> What about all our tables that have competition_id or competitionId as a foreign key? Are they all safe from this change?

Yes, I didn't touch the `id` column and every foreign key is untouched as well.

> What happens when the Results Team uploads results from a competition whose slug != it's id? I think this depends on if we change our public export at all: right now, the WCA Workbook Assistant will generate JSON with the id, which would be the right thing to do (we want our Results to point to a competition's id, not its slug, right?)

Everything non-human-related (automatic tools etc.) should always use the `id`. `slug` is used only for the URL at the moment.

> This is actually pretty hard for me to reason about =( Right now, I'm feeling we should wait until we're ready to go the whole nine yards and change the id to an autoincrementing integer, sorry. What's the motivation for pushing this half solution live now, instead of waiting to go the distance?

How would changing the column to autoincrementing integers help you reason about it? Unless we use a statically typed language where we can enforce that no one passes a slug as an id we're always one step away from doing some stupid stuff like `result.competitionId = c.slug` (which ActiveRecord/MySQL happily accepts).

> What's the motivation for pushing this half solution live now, instead of waiting to go the distance?

Not wasting time with merge conflicts would be my personal reason. I wouldn't call it "half", btw. It's only missing a migration (`varchar -> int`) in order to be a "full" solution. As mentioned above: That migration wouldn't change any code (except for removing the remaining `self.id = ` in the Competition model) and it would be only visible if you look at the raw data/JSON output etc.

We can also approach it the other way around: Get rid of the `id` field in the frontend and prevent people from changing the id. I'm not sure why they need to in the first place.

jfly commented:
My source of confusion/uncertaintly doesn't come from our own code, I'm nervous about what's going to happen when it crosses boundaries to other software/websites not written by us. I'm worried that we're going to generate a lot of confusion among delegates and the results team. You make good points, and maybe my worries are groundless. That said, to help me sleep at night, how do you feel about this one additional piece: write a migration that changes all our ids to something that makes it very clear if we ever see one in the wild that it's an id and not a slug? I'm thinking something like "prefix all our ids with "Immutable".

timhabermaas commented:
I see.

> That said, to help me sleep at night, how do you feel about this one additional piece: write a migration that changes all our ids to something that makes it very clear if we ever see one in the wild that it's an id and not a slug?

Actually this would give *me* sleepless nights. ;) I'd be worried to miss a few foreign keys or break existing code which assumes a specific `id` format.
If we do this, may I suggest replacing the ids with UUIDs (represented as Strings)? `"a80de68a-4e16-46bb-90cc-52b634739148"` screams "don't touch me" to me.

jfly commented:
Haha, I like that idea!

With regards to your sleepless nights, I hear you. However, it is true that at some point in time, we plan to change the id field to an autoincrementing id, yes? At that point in time, we'll have to go into all our tables and update the foreign keys (`competitionId` or `competition_id`, AFAIK). As you know, there's also a list in the code you deleted =) See: https://github.com/cubing/worldcubeassociation.org/blob/87ec41a4056dd0f6b3d9e37b2789abffaec2568a/WcaOnRails/app/models/competition.rb#L196-L201.

Granted, I don't know if changing the form of the `id` would break existing code.

How about this: can you write an email to results@ (go ahead and cc software@) and tell them we're going to stop updating `ids`, and present them with 3 options: leave the ids alone (as you've done here), prefix all the ids with "immutable" (as I suggested), or change them all to UUIDs?

pingskills commented:
@timhabermaas @jfly 

Hi Tim,

Great work with PR. I'd love to see this live so all that horrible code regarding updating keys is removed. I have one suggestion that may help. I think we should also make slug immutable and remove it from the form. If anyone changes the slug then any external links will break. Also be not allowing the slug to be changed the id will always equal the slug until we make the id an integer. Hopefully this resolves any concerns Jeremy has? It also makes changing the id to an integer easier later on because all existing URLs will continue to work (because the slug was equal to the id).

Let me know if you think this is a good idea?

jfly commented:
I'd be okay with not supporting changing the id for now, although I believe it is a feature we should add back later. IDs get changed before a competition is public, so there aren't any external links to worry about.

pingskills commented:
Hey @timhabermaas,

As part of issue #95 I'm looking at creating a competition_events table to clean up the eventSpecs. Having this PR promoted would simplify things as I wouldn't have to worry about IDs changing.

Do you need any help to get this PR over the line?

timhabermaas commented:
@pingskills Sorry for not responding earlier. I'll take care of this PR tomorrow.

I'm fine with removing the friendly_id from the form btw.

jfly commented:
Closing due to inactivity. See our new policy: [PRs that have gone 3 months without activity are subject to being closed.](https://github.com/thewca/worldcubeassociation.org/wiki/Merging-and-deploying)
